### ISSUE TITLE
Allow to review pdf instead of download

### ISSUE THREAD

**Issue by GopherJ** [svenstaro/miniserve#1150]

**Comment by svenstaro** [comment]
You mean you'd like to view PDFs directly in the browser instead of having the browser offer you to download the PDF?

**Comment by SF73** [comment]
I don't know if it refers to this, but it would be nice to have `Content-Disposition:inline` for pdf files.
That way, it would be possible to view them without having to download them.

I don't know if the issue should be adressed here or in actix-files [here](https://github.com/actix/actix-web/blob/d8df60bf4c04c3cbb99bcf19a141c202223e07ea/actix-files/src/named.rs#L134)

**Comment by tylerthecoder** [comment]
I would like this too. I am trying `miniserve --header "Content-Disposition:inline"` and I think it is getting overwritten.

**Comment by GopherJ** [comment]
@svenstaro yes, like mozilla's pdf.js, users can choose if they would like to download or not later

**Comment by svenstaro** [comment]
I think the overwriting might be done by actix components. Can someone check out the code path? I'm not sure when I can get to it.

**Comment by GopherJ** [comment]
@svenstaro  @SF73 pointed out https://github.com/actix/actix-web/blob/d8df60bf4c04c3cbb99bcf19a141c202223e07ea/actix-files/src/named.rs#L134

**Comment by GopherJ** [comment]
I applied the following patch to make it worked

```
diff --git a/actix-files/src/named.rs b/actix-files/src/named.rs
index d7795ba7..308cf702 100644
--- a/actix-files/src/named.rs
+++ b/actix-files/src/named.rs
@@ -134,7 +134,9 @@ impl NamedFile {
                 mime::IMAGE | mime::TEXT | mime::AUDIO | mime::VIDEO => DispositionType::Inline,
                 mime::APPLICATION => match ct.subtype() {
                     mime::JAVASCRIPT | mime::JSON => DispositionType::Inline,
-                    name if name == "wasm" || name == "xhtml" => DispositionType::Inline,
+                    name if name == "wasm" || name == "xhtml" || name == "pdf" => {
+                        DispositionType::Inline
+                    }
                     _ => DispositionType::Attachment,
                 },
                 _ => DispositionType::Attachment,

```

**Comment by svenstaro** [comment]
Could you post the patch

**Comment by SF73** [comment]
Hi, just to get back to this issue.

> I applied the following patch in actix-files to make it worked

I didn't check the code in details but it may be cleaner to handle it on miniserve side rather than asking a patch for the default behavior of actix-files.

I'm don't know rust at all but I guess it could be done [here](https://github.com/svenstaro/miniserve/blob/9bd740b2d07254c4e2a7ae255a19350ce6e2e6de/src/listing.rs#L148) using [set_content_disposition](https://docs.rs/actix-files/latest/actix_files/struct.NamedFile.html#method.set_content_disposition).
> Set the Content-Disposition for serving this file. This allows changing the inline/attachment disposition as well as the filename sent to the peer.

I would like to contribute to implement a config to set the disposition based on some rules but it would take me a lot of time as I don't know rust...

**Comment by duskmoon314** [comment]
I just tried @SF73's suggestion and successfully enabled preview of PDF when only serving a file. I haven't found how to set `Content-Disposition` for file entries listed in directories.

My implementation is here: https://github.com/duskmoon314/miniserve/tree/preview_pdf.

**Comment by duskmoon314** [comment]
I reviewed the issue and related code paths. Here's what I found that might move things forward.

For a file in `Files` service, the [`serve_named_file`](https://github.com/actix/actix-web/blob/1f9bc2de7df3fbf7146e37626df2f6c066eb9a56/actix-files/src/service.rs#L65) is called:

```rust
fn serve_named_file(&self, req: ServiceRequest, mut named_file: NamedFile) -> ServiceResponse {
    if let Some(ref mime_override) = self.mime_override {
        let new_disposition = mime_override(&named_file.content_type.type_());
        named_file.content_disposition.disposition = new_disposition;
    }
    named_file.flags = self.file_flags;
    let (req, _) = req.into_parts();
    let res = named_file.into_response(&req);
    ServiceResponse::new(req, res)
```

Here, `mime_override` is a parameter of `Files` that we can configure on the miniserve side. Something like:

```diff
// src/main.rs configure_app
fn configure_app(app: &mut web::ServiceConfig, conf: &MiniserveConfig) {
    let dir_service = || {
...
+       let mime_override = | _name: &mime:Name | {
+           actix_web::http::header::DispositionType::Inline
+       };
        let base_path = conf.path.clone();
        let no_symlinks = conf.no_symlinks;
        files
            .show_files_listing()
            .files_listing_renderer(listing::directory_listing)
            .prefer_utf8(true)
            .redirect_to_slash_directory()
            .path_filter(move |path, _| {
                // deny symlinks if conf.no_symlinks
                !(no_symlinks && base_path.join(path).is_symlink())
            })
+           .mime_override(mime_override)
...
```

However, as shown above, the `mime_override` only takes `named_file.content_type.type_()` as an argument; that is, we could only see `application` instead of `application/pdf` in `mime_override`.

I tried only returning `DispositionType::Inline`, and everything seems fine: PDF and images are previewed in the browser, and the Typst file (I used for testing here) triggers a download window. However, this method may still cause problems since all responses have the inline header.

I think a better solution is to modify `actix_files`'s `mime_override` to get more information. Or maybe we can find out all PDF files and add `NamedFile` services beside `Files`
